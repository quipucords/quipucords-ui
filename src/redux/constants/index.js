import { aboutModalTypes } from './aboutModalConstants';
import { confirmationModalTypes } from './confirmationModalConstants';
import { credentialsTypes } from './credentialsConstants';
import { factsTypes } from './factsConstants';
import { reportsTypes } from './reportsConstants';
import { scansTypes } from './scansConstants';
import { sourcesTypes } from './sourcesConstants';
import { statusTypes } from './statusConstants';
import { toastNotificationTypes } from './toastNotificationConstants';
import { userTypes } from './userConstants';
import { viewTypes } from './viewConstants';

const reduxTypes = {
  aboutModal: aboutModalTypes,
  confirmationModal: confirmationModalTypes,
  credentials: credentialsTypes,
  facts: factsTypes,
  reports: reportsTypes,
  scans: scansTypes,
  sources: sourcesTypes,
  status: statusTypes,
  toastNotifications: toastNotificationTypes,
  user: userTypes,
  view: viewTypes
};

export {
  reduxTypes as default,
  reduxTypes,
  aboutModalTypes,
  confirmationModalTypes,
  credentialsTypes,
  factsTypes,
  reportsTypes,
  scansTypes,
  sourcesTypes,
  statusTypes,
  toastNotificationTypes,
  userTypes,
  viewTypes
};
