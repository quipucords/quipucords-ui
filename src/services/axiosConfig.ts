// src/services/axiosConfig.ts

import axios from 'axios';
import cookies from 'js-cookie';
import helpers from '../helpers';

/**
 * This helper function should live here, outside of any hook.
 */
const getDecodedToken = (): string => {
  const token = cookies.get(`${process.env.REACT_APP_AUTH_COOKIE}`) || '';
  try {
    return window.atob(token);
  } catch (e) {
    if (!helpers.TEST_MODE) {
      console.error('Invalid token, unable to parse format');
    }
    return '';
  }
};

// --- REQUEST INTERCEPTOR ---
// This runs BEFORE every API request.
axios.interceptors.request.use(config => {
  const token = getDecodedToken();
  if (token) {
    config.headers.Authorization = `Token ${token}`;
  }
  return config;
});

// --- RESPONSE INTERCEPTOR ---
// This runs on every API response.
axios.interceptors.response.use(
  response => response, // Do nothing on a successful response
  error => {
    // On a 401 error, remove the cookie and force a reload to the login page.
    if (error?.response?.status === 401) {
      cookies.remove(`${process.env.REACT_APP_AUTH_COOKIE}`);
      document.location.replace('./');
    }
    return Promise.reject(error);
  }
);

/**
 * We export a function that can be used to check the initial auth state,
 * but the interceptors are already active just by importing this file.
 */
export const hasInitialAuthToken = (): boolean => {
  return !!getDecodedToken();
};
