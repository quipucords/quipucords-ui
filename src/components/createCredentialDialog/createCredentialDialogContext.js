import { useState } from 'react';
import { AlertVariant } from '@patternfly/react-core';
import { useShallowCompareEffect } from 'react-use';
import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import { useInferredContext as useViewContext } from '../view/viewContext';
import { apiTypes } from '../../constants/apiConstants';
import { translate } from '../i18n/i18n';

/**
 * Return credential dialog response
 *
 * @param {object} options
 * @param {Function} options.useSelector
 * @returns {*}
 */
const useCredential = ({ useSelector: useAliasSelector = storeHooks.reactRedux.useSelector } = {}) => {
  const { credentialType, ...data } = useAliasSelector(({ credentials }) => credentials?.dialog, {});
  const updatedCredentialType = data?.credential?.[apiTypes.API_QUERY_TYPES.CREDENTIAL_TYPE] || credentialType;

  return {
    credentialType: updatedCredentialType,
    ...data
  };
};

/**
 * Return an updated source. Display relative toast messaging after wizard closes.
 *
 * @param {object} options
 * @param {Function} options.addCredential
 * @param {Function} options.getCredentials
 * @param {Function} options.t
 * @param {Function} options.updateCredential
 * @param {Function} options.useCredential
 * @param {Function} options.useDispatch
 * @param {Function} options.useViewContext
 * @returns {(function(*, *): void)|*}
 */
const useOnSubmitCredential = ({
  addCredential = reduxActions.credentials.addCredential,
  getCredentials = reduxActions.credentials.getCredentials,
  t = translate,
  updateCredential = reduxActions.credentials.updateCredential,
  useCredential: useAliasCredential = useCredential,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useViewContext: useAliasViewContext = useViewContext
} = {}) => {
  const { viewId } = useAliasViewContext();
  const [credential, setCredential] = useState();
  const dispatch = useAliasDispatch();
  const { edit, fulfilled } = useAliasCredential();

  useShallowCompareEffect(() => {
    if (credential?.values) {
      const { id, ...data } = credential.values;

      if (id) {
        updateCredential(id, data)(dispatch);
      } else {
        addCredential(data)(dispatch);
      }
    }
  }, [addCredential, credential, dispatch, updateCredential]);

  useShallowCompareEffect(() => {
    if (credential?.values && fulfilled) {
      dispatch([
        {
          type: reduxTypes.credentials.UPDATE_CREDENTIAL_HIDE
        },
        {
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: AlertVariant.success,
          header: t('toast-notifications.title', { context: ['credential', edit && 'edit'] }),
          message: t('toast-notifications.description', { context: ['credential', edit && 'edit'] })
        },
        {
          type: reduxTypes.credentials.RESET_ACTIONS
        },
        {
          type: reduxTypes.view.UPDATE_VIEW,
          viewId
        }
      ]);

      getCredentials()(dispatch);
    }
  }, [credential, dispatch, edit, fulfilled, getCredentials, t]);

  return data => setCredential(() => data);
};

/**
 * Functions for handling state updates for credentials.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {{onEdit: Function, onHide: Function, onAdd: Function}}
 */
const useOnUpdateCredential = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  /**
   * On edit show the credential in dialog
   *
   * @event onEdit
   * @param {object|*} credential
   */
  const onEdit = credential => {
    dispatch({
      type: reduxTypes.credentials.EDIT_CREDENTIAL_SHOW,
      credential
    });
  };

  /**
   * Hide the credential dialog
   *
   * @event onHide
   */
  const onHide = () => {
    dispatch([
      {
        type: reduxTypes.credentials.UPDATE_CREDENTIAL_HIDE
      },
      {
        type: reduxTypes.credentials.RESET_ACTIONS
      }
    ]);
  };

  /**
   * on add show the credential dialog
   *
   * @event onAdd
   * @param {string|*} credentialType
   */
  const onAdd = credentialType => {
    dispatch([
      {
        type: reduxTypes.credentials.CREATE_CREDENTIAL_SHOW,
        credentialType
      }
    ]);
  };

  return {
    onEdit,
    onHide,
    onAdd
  };
};

const context = {
  useCredential,
  useOnSubmitCredential,
  useOnUpdateCredential
};

export { context as default, context, useCredential, useOnSubmitCredential, useOnUpdateCredential };
