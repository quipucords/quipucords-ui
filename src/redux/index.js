import React from 'react';
import { connect } from 'react-redux';
import { useLocation, useNavigate, useParams } from 'react-router-dom';
import { store } from './store';
import { reduxActions } from './actions';
import { reduxHelpers } from './common';
import { storeHooks } from './hooks';
import { reduxReducers } from './reducers';
import { reduxSelectors } from './selectors';
import { reduxTypes } from './constants';

/**
 * Patch withRouter for legacy components.
 *
 * @param {React.ReactNode} Component
 * @returns {React.ReactNode}
 */
const withRouter = Component => props => {
  const location = useLocation();
  const navigate = useNavigate();
  const params = useParams();
  return (
    <Component
      {...props}
      history={{ push: navigate }}
      location={location}
      navigate={navigate}
      params={params}
      router={{ location, navigate, params }}
    />
  );
};

const connectRouter = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(connect(mapStateToProps, mapDispatchToProps)(component));

export {
  connect,
  connectRouter,
  reduxActions,
  reduxHelpers,
  reduxReducers,
  reduxSelectors,
  reduxTypes,
  store,
  storeHooks
};
