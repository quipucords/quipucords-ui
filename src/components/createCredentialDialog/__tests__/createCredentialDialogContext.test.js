import React from 'react';
import { context, useOnSubmitCredential, useOnUpdateCredential } from '../createCredentialDialogContext';
import { store } from '../../../redux';
import { credentialsService } from '../../../services';

describe('CredentialsContext', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should attempt to handle submitting a credential with a toast confirmation', async () => {
    const mockCredential = {
      values: {
        mockId: 'lorem ipsum base id',
        mockName: 'lorem ipsum name'
      }
    };
    const mockUseState = jest.spyOn(React, 'useState').mockImplementation(() => [mockCredential, jest.fn()]);
    const mockUseService = jest.spyOn(credentialsService, 'getCredentials').mockImplementation(() => jest.fn());

    const { result } = await mountHook(() =>
      useOnSubmitCredential({
        useDispatch: () => mockDispatch,
        useCredential: () => ({ fulfilled: true })
      })
    );

    result(mockCredential);

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch onSubmit');
    mockUseState.mockClear();
    mockUseService.mockClear();
  });

  it('should handle credential actions for onEdit, onHide, onAdd', async () => {
    const { result } = await shallowHook(() =>
      useOnUpdateCredential({
        useDispatch: () => mockDispatch
      })
    );

    result.onEdit({
      mockId: 'lorem ipsum base id',
      mockName: 'lorem ipsum name'
    });
    result.onHide();
    result.onAdd('lorem ipsum mock credential type');

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch onEdit');
  });
});
