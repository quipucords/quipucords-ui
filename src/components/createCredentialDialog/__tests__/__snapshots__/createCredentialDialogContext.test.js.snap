// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialsContext should attempt to handle submitting a credential with a toast confirmation: dispatch onSubmit 1`] = `
Array [
  Array [
    Object {
      "payload": Promise {},
      "type": "ADD_CREDENTIAL",
    },
  ],
  Array [
    Array [
      Object {
        "type": "UPDATE_CREDENTIAL_HIDE",
      },
      Object {
        "alertType": "success",
        "header": "t(toast-notifications.title, {\\"context\\":\\"credential\\"})",
        "message": "t(toast-notifications.description, {\\"context\\":\\"credential\\"})",
        "type": "TOAST_ADD",
      },
      Object {
        "type": "RESET_CREDENTIAL_ACTIONS",
      },
      Object {
        "type": "UPDATE_VIEW",
        "viewId": undefined,
      },
    ],
  ],
  Array [
    Object {
      "payload": [MockFunction],
      "type": "GET_CREDENTIALS",
    },
  ],
]
`;

exports[`CredentialsContext should handle credential actions for onEdit, onHide, onAdd: dispatch onEdit 1`] = `
Array [
  Array [
    Object {
      "credential": Object {
        "mockId": "lorem ipsum base id",
        "mockName": "lorem ipsum name",
      },
      "type": "EDIT_CREDENTIAL_SHOW",
    },
  ],
  Array [
    Array [
      Object {
        "type": "UPDATE_CREDENTIAL_HIDE",
      },
      Object {
        "type": "RESET_CREDENTIAL_ACTIONS",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "credentialType": "lorem ipsum mock credential type",
        "type": "CREATE_CREDENTIAL_SHOW",
      },
    ],
  ],
]
`;

exports[`CredentialsContext should return specific properties: specific properties 1`] = `
Object {
  "useCredential": [Function],
  "useOnSubmitCredential": [Function],
  "useOnUpdateCredential": [Function],
}
`;
